A blog post about jenkins cli (#2389)
jenkins-infra/jenkins.io: content/blog/2019/07/2019-07-30-jenkins-cli.adoc

---
layout: post
title: "Managing Jenkins with jcli"
tags:
- general
- jcli
author: linuxsuren
---

As a developer, I usually use Jenkins like this:

> * Find a job which is related with your current work

Reviewer: ```suggestion
* Find a job which is related with my current work
```
... or change "I" to "you" in the whole paragraph

Visit Jenkins page from `http://localhost:8080/jenkins/cli/`. You'll see a command like `java -jar jenkins-cli.jar -s http://localhost:8080/jenkins/ help`. So, a jar file needs to be download.
We can use this command to complete this task `wget http://localhost:8080/jenkins/jnlpJars/jenkins-cli.jar`.

Now you can see that this is not a Linux-style CLI. Please consider some points below:

* The users must have a JRE. This is not convenient for developers who don't use Java.
* Type more words. We always need to type `java -jar jenkins-cli.jar -s http://localhost:8080/jenkins/` as the initial command.
* Cannot install it by some popular package manager, like `brew` or `yum`.

Of course, the Java CLI client is more native with Jenkins. But I'd like to use this more easily.

That's the story of creating `jcli`.

== Features

* Easy to maintain config file for jcli
* Multiple Jenkins support
* Plugins management (list, search, install, upload)
* Job management (search, build, log)
* Open your Jenkins with a browser
* Restart your Jenkins
* Connection with proxy support

== How to get it?

> You can find `jcli` from this repo https://github.com/jenkins-zh/jenkins-cli/[jenkins-cli]. For now, we support 

Reviewer: ```suggestion
You can clone `jcli` from https://github.com/jenkins-zh/jenkins-cli/[the jenkins-cli repo]. For now, we support 
```

these three most popular OS platforms.

=== MacOS

You can use brew to install jcli.

```
brew tap jenkins-zh/jcli
brew install jcli
```

=== Linux

It's very simple to install jcli into your Linux OS. Just need to execute a command line at below:

```
curl -L https://github.com/jenkins-zh/jenkins-cli/releases/latest/download/jcli-linux-amd64.tar.gz|tar xzv
sudo mv jcli /usr/local/bin/
```

=== Windows

You can find the latest version by clicking https://github.com/jenkins-zh/jenkins-cli/releases/latest/download/jcli-windows-386.tar.gz[here]. Then download the tar file, cp the uncompressed `jcli` directory into your system path.

=== How to get started?

> It's very simple to use this. Once you get `jcli` on your computer, use this command to generate and modify your Jenkins configuration:

Reviewer: Maybe just mention that the command generates a sample config here, and describe the suggested modifications in the next paragraph (below the command).

```
jcli config generate
```

This cmd will help to you generate a sample config file. In most cases, you should modify the `url`, `username` and `token`. OK, I believe you're ready. Please check whether you install the `github` plugin in your Jenkins:

`jcli plugin list --filter name=github`

> At the end. It's still under very early develope time. Any contribution is welcome.

Reviewer: ```suggestion
That's the end. It's still in very early development stage. Any contribution is welcome.
```

---
layout: post
title: "Managing Jenkins with jcli"
tags:
- general
- jcli
author: linuxsuren
---

As a developer, I usually use Jenkins like this:

* Find a job which is related with your current work
* Trigger that job
* Check the output of log

> Sometimes, I might need to check the update center. Maybe a new plugin is needed, or update an exists plugin.

Reviewer 2: "or I need to update an existing plugin."

> Or, I want to upload a plugin from my computer. For all these cases, I just don't need a UI or even a browse.

Reviewer 2: "browser."

I like to use a CLI to complete most of my tasks. For example, I use `kubectl` to manage my Kubernetes cluster,

> to create or modify the kubernetes resources. So, I start to think that why not using a CLI to manage my Jenkins.

Reviewer 2: "I start to think, 'Why not use a CLI to manage my Jenkins?'"

== Why create a new one?

> First, I almost forget the Jenkins CLI is existed which is written by Java. Let me do dome introduce how to use the exist one.

Reviewer 2: "First, I almost forgot about the existing Jenkins CLI, written in Java. Let me introduce how to use that one."

Visit Jenkins page from `http://localhost:8080/jenkins/cli/`. You'll see a command like `java -jar jenkins-cli.jar -s http://localhost:8080/jenkins/ help`. So, a jar file needs to be download.
We can use this command to complete this task `wget http://localhost:8080/jenkins/jnlpJars/jenkins-cli.jar`.

> Now, I believe you can see this's not a linux cli style. Please consider some points below:

Reviewer 2: "Now you can see that this is not a Linux-style CLI."

> * The users must have a JRE. It's not convenience for other developers who don't use Java.

Reviewer 2: "This is not convenient for developers who don't use Java."

> * Type more words. We always need to type `java -jar jenkins-cli.jar -s http://localhost:8080/jenkins/` as the inital cmd.

Reviewer 2: "as the initial command."

* Cannot install it by some popular package manager, like `brew` or `yum`.

> Of course, the Java CLI client is more native with Jenkins. But I'd like to use this more easier.

Reviewer 2: "more easily."

That's the story of creating `jcli`.

== Features

* Easy to maintain config file for jcli
* Multiple Jenkins support
* Plugins management (list, search, install, upload)
* Job management (search, build, log)

> * Open your Jenkins with a browse

Reviewer 2: "browser"

* Restart your Jenkins
* Connection with proxy support

== How to get it?

You can find `jcli` from this repo https://github.com/jenkins-zh/jenkins-cli/[jenkins-cli]. For now, we support 

> these three most popular os platform.

Reviewer 2: "OS platforms."

=== MacOS

You can use brew to install jcli.

```
brew tap jenkins-zh/jcli
brew install jcli
```

=== Linux

It's very simple to install jcli into your Linux OS. Just need to execute a command line at below:

```
curl -L https://github.com/jenkins-zh/jenkins-cli/releases/latest/download/jcli-linux-amd64.tar.gz|tar xzv
sudo mv jcli /usr/local/bin/
```

=== Windows

> You can find the latest version by click https://github.com/jenkins-zh/jenkins-cli/releases/latest/download/jcli-windows-386.tar.gz[here]. Then download the tar file, cp the uncompressed `jcli` directory into your system path.

Reviewer 2: "clicking"

=== How to get started?

> It's very simple to use this. Once you got `jcli` in your computer. One thing is needed, use below cmd to generate and 

Reviewer 2: "Once you get jcli on your computer, use this command to generate and modify your Jenkins configuration:"

Sometimes, I might need to check the update center. Maybe a new plugin is needed, or I need to update an existing plugin.
Or, I want to upload a plugin from my computer. For all these cases, I just don't need a UI or even a browser.
I like to use a CLI to complete most of my tasks. For example, I use `kubectl` to manage my Kubernetes cluster,
to create or modify the kubernetes resources. So, I start to think, 'Why not use a CLI to manage my Jenkins?'.

== Why create a new one?

First, I almost forgot about the existing Jenkins CLI, written in Java. Let me introduce how to use that one.

Visit Jenkins page from `http://localhost:8080/jenkins/cli/`. You'll see a command like `java -jar jenkins-cli.jar -s http://localhost:8080/jenkins/ help`. So, a jar file needs to be download.
We can use this command to complete this task `wget http://localhost:8080/jenkins/jnlpJars/jenkins-cli.jar`.

Now you can see that this is not a Linux-style CLI. Please consider some points below:

* The users must have a JRE. This is not convenient for developers who don't use Java.
* Type more words. We always need to type `java -jar jenkins-cli.jar -s http://localhost:8080/jenkins/` as the initial command.

> * Cannot install it by some popular package manager, like `brew` or `yum`.

Reviewer 3: Also the startup time is slow. Not sure if that was a factor for _you_.

Author: Thanks for your reminder. I'd like to move to next week.

It's very simple to install jcli into your Linux OS. Just need to execute a command line at below:

```

> curl -L https://github.com/jenkins-zh/jenkins-cli/releases/latest/download/jcli-linux-amd64.tar.gz|tar xzv

Reviewer 3: BTW snapcraft.io would be great

Author: Yes. I've created an issue for this. Try to support snapcraft later. https://github.com/jenkins-zh/jenkins-cli/issues/121
